### PR TITLE
feature: add resource detail view

### PR DESCRIPTION
**Pull request by zaharidichev**
This PR introduces the resoruce detail view. It also brings a couple of additional changes: 

- we now rely on context of entered spans rather than explicit parent - child relationships
- we expose internal resources (BatchSempahore)
- we show the relationships between resources (Mutex -> BatchSemaphore)
- we show the async ops that are live on a specific resource
- we also show the tasks that are awaiting on these async ops, which allows the user to draw a relationship between tasks and resources
- there are more examples added in the examples folder (semaphore and mutex)

Some screenshots: 

Screenshot 2021-11-23 at 19 12 28
Screenshot 2021-11-23 at 19 12 38

Screenshot 2021-11-23 at 19 12 50

Signed-off-by: Zahari Dichev <zaharidichev@gmail.com>

### CODE REVIEW

**Comment by hawkw** [comment]
This looks _amazing_. I'm going to try to give this branch a detailed review today, but the screenshots look awesome. Amazing work!

**Comment by zaharidichev** [comment]
@hawkw Thanks a bunch for taking the time. Let me know after a quick glance whether you prefer me to break up the PR somehow or whether I can do anything to make it easier to review.

File: console-api/proto/async_ops.proto
Comment (hawkw) [inline console-api/proto/async_ops.proto:44]: nit: it might be nice if this comment explained a bit more what the "parent async op" of an async op is --- my guess is that this is for situations such as a `Sleep` inside of an `Interval`, or a `Semaphore` op inside a channel? but, it would be nice if the comment explained that.

also, if a given async op doesn't have a parent, this will not be set, right? it might be worth documenting that as well.

File: console-api/proto/common.proto
Comment (hawkw) [inline console-api/proto/common.proto]: nit:
```suggestion
// State attributes of an entity. These are dependent on the type of the entity.
//
// For example, a timer resource will have a duration, while a semaphore resource may
```

File: console-api/proto/common.proto
Comment (hawkw) [inline console-api/proto/common.proto]: nit/typo:
```suggestion
// have permits as an attribute. Likewise, the async ops of a semaphore may have attributes
```

File: console-subscriber/examples/barrier.rs
Comment (hawkw) [inline console-subscriber/examples/barrier.rs:1]: thanks for adding new examples, these are great!

take it or leave it, but it might also be nice to have some examples that combine multiple resource types. that way, we could simulate what a real application that uses a large number of resources might look like. we could maybe stick similar code to these examples into `examples/app.rs` (maybe with a CLI option) so that the example can also spawn a number of tasks that use barriers, mutexen, etc.

it would be totally fine to make that change in a follow-up branch, though.

File: console-subscriber/examples/mutex.rs
Comment (hawkw) [inline console-subscriber/examples/mutex.rs]: nit, take it or leave it: this seems like it could be a `while` loop:
```suggestion
            while *count.lock().await < 50 {}
```

File: console-subscriber/src/aggregator/mod.rs
Comment (hawkw) [inline console-subscriber/src/aggregator/mod.rs]: nit: i would probably import `std::collections::hash_map` rather than spelling this out. not a blocker, though.

File: console-subscriber/src/aggregator/mod.rs
Comment (hawkw) [inline console-subscriber/src/aggregator/mod.rs]: nit, take it or leave it: i think we could combine this into one `and_then` like:
```suggestion
                            .and_then(|r| self.resources.get(r.parent_id.as_ref()?))
```

File: console-subscriber/src/aggregator/mod.rs
Comment (hawkw) [inline console-subscriber/src/aggregator/mod.rs]: similarly, i think this could be
```suggestion
nit, take it or leave it: i think we could combine this into one `and_then` like:
```suggestion
                            .and_then(|r| self.async_ops.get(r.parent_id.as_ref()?))
```
```

File: console-subscriber/src/aggregator/mod.rs
Comment (hawkw) [inline console-subscriber/src/aggregator/mod.rs]: take it or leave it: this could also be written as:
```suggestion
                    e
                        .and_modify(|attr| update_attribute(attr.get_mut(), upd))
                        .or_insert_with(|| upd.clone().into());
```

File: console-subscriber/src/lib.rs
Comment (hawkw) [inline console-subscriber/src/lib.rs]: at this point we care about enough different sets of callsites that i wonder if it might be worth having one array of call the callsites that stores what kind of callsite it is...but, we would need to look at the current usage patterns to determine whether that would be better or worse. anyway, probably not something we want to address in this PR.

File: console-subscriber/src/lib.rs
Comment (hawkw) [inline console-subscriber/src/lib.rs:157]: it might be worth adding comments explaining what some of these fields mean, for future readers? not a blocker though.

File: console-subscriber/src/lib.rs
Comment (hawkw) [inline console-subscriber/src/lib.rs]: i'm not the _biggest_ fan of this naming..."target" has a specific meaning in `tracing` terminology, and technically the `TargetType` enum represents the event's `tracing` target. what about maybe just calling this `StateUpdate`/`StateUpdateType`?

File: console-subscriber/src/lib.rs
Comment (hawkw) [inline console-subscriber/src/lib.rs]: that...certainly is some formatting...

i wonder if this `if` chain could be made a little nicer using early returns? not a blocker though.

File: console-subscriber/src/lib.rs
Comment (hawkw) [reply console-subscriber/src/lib.rs]: also, AFAICT, the `Target` struct seems to only be used as a field on `StateUpdate`...is it really necessary for this to be a separate struct, or can its fields just be part of `StateUpdate`?

File: console-subscriber/src/visitors.rs
Comment (hawkw) [inline console-subscriber/src/visitors.rs]: this is a _big_ tuple...at this point, it really feels like maybe this should be returning a struct with named fields instead?

File: console-subscriber/src/lib.rs
Comment (hawkw) [inline console-subscriber/src/lib.rs:519]: minor nit, something we can always do in a follow-up: it would be nice if we could find both the task ID _and_ the async op ID in one iteration over the stack, instead of iterating twice...these stacks are generally not going to be so deeply nested that this is _really_ important, but it would be better to only iterate once, i think...

File: console-subscriber/src/visitors.rs
Comment (hawkw) [inline console-subscriber/src/visitors.rs:34]: hmm, the fact that we have a sort of magic field that controls how the console displays data is...not exactly my favorite thing? it seems like in an ideal world, the tracing events would _just_ expose data, and the console would be able to determine how it should be displayed, and this kind of feels like a violation of that idea. but, i also get that maybe this is necessary to get the data to display nicely.

is there any other potential solution that would let us get the UI we want without having the runtime annotate its events with this field? because...that would certainly be nice.

File: console-subscriber/src/visitors.rs
Comment (hawkw) [inline console-subscriber/src/visitors.rs]: this tuple feels big enough that it should _probably_ be a struct...

File: console/src/state/async_ops.rs
Comment (hawkw) [inline console/src/state/async_ops.rs:1]: this file looks fine overall, but also, it does occur to me that we're really starting to reach the point where there's a _lot_ of very similar boilerplate for different parts of the data model (tasks, resources, async ops...). it would maybe be worth thinking about whether some of this code, like the sorting by fields, could be made generic so we don't have to write identical boilerplate quite so many times.

but, that's not really a problem for this branch, just something we might want to think about addressing later...

File: console/src/state/mod.rs
Comment (hawkw) [inline console/src/state/mod.rs:436]: we may want to think about only re-formatting this string when the field value _changes_, rather than every time we render a screen...

File: console/src/view/async_ops.rs
Comment (hawkw) [inline console/src/view/async_ops.rs]: it weould be nice to avoid re-formatting and allocating these strings every time...

File: console/src/view/async_ops.rs
Comment (hawkw) [inline console/src/view/async_ops.rs]: similarly, these could probably be formatted once, i imagine?

File: console/src/view/resource.rs
Comment (hawkw) [inline console/src/view/resource.rs]: ids, targets, and concrete types aren't going to change, so it would be nice to be able to avoid formatting these strings every time.

File: console/src/view/resource.rs
Comment (hawkw) [inline console/src/view/resource.rs]: it would be nice not to format this on every frame...

File: console/src/view/resource.rs
Comment (hawkw) [inline console/src/view/resource.rs]: can we store the kind string as well?

File: console/src/view/resource.rs
Comment (hawkw) [inline console/src/view/resource.rs]: ...and the concrete type string?

File: console/src/view/resources.rs
Comment (hawkw) [inline console/src/view/resources.rs]: i have a slight preference for spelling this "Vis", as it's short for "Visibility"...or maybe it should just be "pub"? but, i could be convinced otherwise...

File: console/src/view/resources.rs
Comment (hawkw) [inline console/src/view/resources.rs]: another string that shouldn't have to be `format!`ed every time we render a frame...

**Review by hawkw — COMMENTED** [review]
overall, this seems pretty good --- and the UI screenshots you posted are awesome! i had a bunch of minor suggestions.

also, i wonder if it would be worth adding a way to hide/show all the internal resources---and maybe hide them by default? i don't know if we _really_ want to expose tokio internals like the batch semaphore by default, unless the user _asks_ for them, it seems like it could potentially just be a bunch of information that's confusing to a user who isn't familiar with tokio's internals. but, we could address that in a separate PR.

File: console-subscriber/src/visitors.rs
Comment (zaharidichev) [reply console-subscriber/src/visitors.rs:34]: @hawkw If you are referring to `inherits_child_attrs` I totally agree. The problem is that in some scenarios we want to propagate these attrs to the parent resource/async op. For example in the situation of Semaphore/BatchSempahore. Most of the instrumentation points that update state are happening in the BatchSempahore but we want the same data on the public Semaphore as it will be a bit hard to instrument that one properly. One possible solution that will most likely yield a similar result to what we have at the moment is to make these state attrs always be inherited from child resources that are flagged as internal. In that scenario however, we will have the `Mutex` resource inherit the state attrs of the internal Semaphore that it uses. Of course we can think of some way to differentiate them in the UI to make it clear that these are attrs of an internal resource but again... this can become confusing. Do you have any better ideas?

Comment (hawkw) [reply console-subscriber/src/visitors.rs:34]: > The problem is that in some scenarios we want to propagate these attrs to the parent resource/async op. For example in the situation of Semaphore/BatchSempahore. Most of the instrumentation points that update state are happening in the BatchSempahore but we want the same data on the public Semaphore as it will be a bit hard to instrument that one properly. 

yeah, i get the motivation for this, and it is entirely possible that it _is_ the best solution! 

one thing i might consider trying is a global toggle in the resources view that sets whether or not private resources are displayed. when private resources are turned on, we would show those resources in the resource list _and_ display any child attributes from private resources on the parent resources. when the toggle is turned off, we would hide both the private resources from the resource list, _and_ hide their attributes from parent resources. it could be worth playing with that and seeing if it's an acceptable solution to this problem?

if a toggle in the UI doesn't adequately solve the problem, i'd be okay with moving forward with this approach; it just seems worth trying?

Comment (hawkw) [reply console-subscriber/src/visitors.rs:34]: another thing to consider is, maybe we just always have parents inherit child attributes if they have no attributes of their own, regardless of whether or not the toggle is set?

File: console/src/view/resource.rs
Comment (zaharidichev) [reply console/src/view/resource.rs]: In these situations do you mean storing the string as an `InternetStr` on the `Resource/AsyncOp` objects?

Comment (hawkw) [reply console/src/view/resource.rs]: Depending on the string, we may or may not want to intern them. In the case of `resource.kind()`, I believe they already _are_ `InternedStr`s: https://github.com/tokio-rs/console/blob/042c1f78447839bed54538ed89b9277e437ed6b4/console/src/state/resources.rs#L33-L37

I think we could probably add a method to `resources::Kind` that returns an `&'static str`, so that we don't need to `format!` it here. I can open a quick PR for that change?

Comment (hawkw) [reply console/src/view/resource.rs]: ~~...actually, i just double-checked, and `resource.kind()` _already_ returns an `&str`. so i think this can just be~~
```suggestion
            Span::raw(resource.kind()),
```


**Edit**: oh, hmm, that doesn't actually work due to lifetime issues. i think we may need to make some additional changes to `InternedStr` to make this work nicely...

File: console-api/proto/async_ops.proto
Comment (hawkw) [inline console-api/proto/async_ops.proto]: some minor nits here:
```suggestion
    // The ID of the parent async op. 
    //
    // This field is only set if this async op was created while inside of another
    // async op.  For example, `tokio::sync`'s `Mutex::lock` internally calls 
    // `Semaphore::acquire`. 
    // 
    // This field can be empty; if it is empty, this async op is not a child of another
    // async op.
```

File: console-api/proto/common.proto
Comment (hawkw) [inline console-api/proto/common.proto]:
```suggestion
// have a permit count. Likewise, the async ops of a semaphore may have attributes
```

File: console-subscriber/src/aggregator/mod.rs
Comment (hawkw) [inline console-subscriber/src/aggregator/mod.rs:808]: i think in theory there might be a way to not repeat this code quite as much, but i don't know if it's worth the effort of figuring that out. we could maybe have a function that takes a hash map and performs the lookup part? but, not a blocker.

File: console/src/state/async_ops.rs
Comment (hawkw) [inline console/src/state/async_ops.rs:37]: i think the parent ID could probably just be a normal `String` rather than an `InternedStr`, unless we expect a single async op to have a _lot_  of children. but, we'll clean up unused interned strings anyway so this is fine in either case.

File: console/src/state/resources.rs
Comment (hawkw) [inline console/src/state/resources.rs]: nit, take it or leave it: this could maybe be simplified to
```suggestion
            let parent_id = strings.string(parent_id
                .map(u64::to_string)
                .unwrap_or_else(|| "n/a".to_string());
            );
```

File: console/src/state/resources.rs
Comment (hawkw) [inline console/src/state/resources.rs]: nit: this should probably be 
```suggestion
                    .unwrap_or_else(|| format!("{}", id)),
```
so that we don't format the string if we _did_ find a parent (see the documentation for https://doc.rust-lang.org/stable/core/option/enum.Option.html#method.unwrap_or)

Comment (hawkw) [reply console/src/state/resources.rs]: also, it might be _slightly_ more efficient not to go through the `format!` macro when the format arguments are just `"{}"`; we can use the `fmt::Display` implementation for `u64` directly:
```suggestion
                    .unwrap_or_else(|| id.to_string()),
```

but, that's probably not a _huge_ performance impact, regardless.

File: console/src/state/async_ops.rs
Comment (hawkw) [inline console/src/state/async_ops.rs]: nit: could be 
```suggestion
                .map(u64::to_string)
```
which might be very slightly more efficient (but not a big deal)

File: console/src/view/async_ops.rs
Comment (hawkw) [inline console/src/view/async_ops.rs]: nit/tioli: could just be
```suggestion
                    let task_str = task.unwrap_or_else(||  async_op.task_id_str().to_owned());
```

**Review by hawkw — APPROVED** [review]
overall, this looks pretty good to me and i think we should merge this and address the bigger potential improvements in follow-up branches. i commented on a few minor suggestions we may want to address first.

File: console/src/state/async_ops.rs
Comment (zaharidichev) [reply console/src/state/async_ops.rs:37]: @hawkw I actually wonder whether we can take a different approach with all the parent info. Namely, what if whenever we have a parent we simply have an `Option<AsyncOpRef> `or  `Option<ResourceRef>`  and then get all the needed strings, etc by upgrading the parent ref? That will make it much easier when we have a toggle that determines whether we should display any info about internal resources. WDYT?

File: console-subscriber/src/aggregator/mod.rs
Comment (pnkfelix) [reply console-subscriber/src/aggregator/mod.rs:808]: Alternatively, you could have a method that takes an `UpdateType` and gives you back the hashmap, and then leave the currently duplicated code inline as a single non-duplicated `if` statement. (This works in part because you can just pass along the `UpdateType` into the `to_update.push` invocation, rather than writing out the specific variant the way you are doing here.

File: console/src/state/async_ops.rs
Comment (hawkw) [reply console/src/state/async_ops.rs:37]: We could definitely do something like that; it might be worth doing in a follow-up branch so that we can get this PR merged sooner?

File: console-subscriber/src/aggregator/mod.rs
Comment (hawkw) [reply console-subscriber/src/aggregator/mod.rs:808]: yeah, that's more or less what i was imagining. i think the `filter` might have to be kept separate, as well, because the field access is being performed on a different type. we could introduce a trait for that, but, hmm...

File: console/src/state/async_ops.rs
Comment (zaharidichev) [reply console/src/state/async_ops.rs:37]: 👍

**Comment by hawkw** [comment]
@zaharidichev is this ready to merge or are you planning on making an additional changes in this branch?

**Comment by zaharidichev** [comment]
I can address more of the feedback in follow-ups. Its already getting quite big.  I think the corresponding tokio branch needs to be sanity checked before we merge this though:  https://github.com/tokio-rs/tokio/pull/4302

**Comment by carllerche** [comment]
@hawkw @zaharidichev to confirm: this PR is ready to merge pending the Tokio PR?

**Comment by carllerche** [comment]
@zaharidichev Tokio 1.15 is out.

File: console-api/proto/common.proto
Comment (hawkw) [inline console-api/proto/common.proto]: just adding the docs addition from @pnkfelix's PR:
```suggestion
    message Attribute {
        // The key-value pair for the attribute
        common.Field field = 1;
        // Some values carry a unit of measurement. For example, a duration
        // carries an associated unit of time, such as "ms" for milliseconds.
        optional string unit = 2;
    }
```

File: console-api/proto/resources.proto
Comment (hawkw) [inline console-api/proto/resources.proto:50]:
```suggestion
    // Is the resource an internal component of another resource?
    //
    // For example, a `tokio::time::Interval` resource might contain a 
    // `tokio::time::Sleep` resource internally.
    bool is_internal = 7;
```

**Review by hawkw — APPROVED** [review]
added docs

File: console-api/proto/common.proto
Comment (hawkw) [inline console-api/proto/common.proto]: oh whoops my copy-paste indented this too much:
```suggestion
message Attribute {
    // The key-value pair for the attribute
    common.Field field = 1;
    // Some values carry a unit of measurement. For example, a duration
    // carries an associated unit of time, such as "ms" for milliseconds.
    optional string unit = 2;
}
```

**Pull request merged**
